src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+            if (waterTerrs == null) {
+              final LinkedHashSet<String> newWaterTerrs = new LinkedHashSet<String>();
+              newWaterTerrs.add(waterTerr);
+              landWaterTerritoyConnections.put(landTerr, newWaterTerrs);
+            } else
+              waterTerrs.add(waterTerr);
+          }
+
+        }
+        final HashMap<String, LinkedHashSet<String>> landWaterTerrConnChecks =
+            new HashMap<String, LinkedHashSet<String>>(landWaterTerritoyConnections);
+        for (final Entry<String, LinkedHashSet<String>> landWaterTerrConn : landWaterTerritoyConnections.entrySet()) {
+          final String landTerr = landWaterTerrConn.getKey();
+          final LinkedHashSet<String> waterTerrs = landWaterTerrConn.getValue();
+          LinkedHashSet<String> landTerrNeighbors = MapXMLHelper.territoryConnections.get(landTerr);
+          if (landTerrNeighbors == null)
+            landTerrNeighbors = new LinkedHashSet<String>();
+          landWaterTerrConnChecks.remove(landTerr);
+          for (final Entry<String, LinkedHashSet<String>> landWaterTerrConn2 : landWaterTerrConnChecks.entrySet()) {
+            final String landTerr2 = landWaterTerrConn2.getKey();
+            // if (landTerrNeighbors.contains(landTerr2))
+            // continue;
+            LinkedHashSet<String> landTerrNeighbors2 = MapXMLHelper.territoryConnections.get(landTerr2);
+            if (landTerrNeighbors2 == null)
+              landTerrNeighbors2 = new LinkedHashSet<String>();
+            // if (landTerrNeighbors2.contains(landTerr))
+            // continue;
+            if (!landTerrNeighbors.contains(landTerr2) && !landTerrNeighbors2.contains(landTerr))
+              continue;
+            final SortedSet<String> waterTerrs2 = new TreeSet<String>(landWaterTerrConn2.getValue());
+            waterTerrs2.retainAll(waterTerrs);
+            if (waterTerrs2.size() > 1) {
+              // remove water territories that are not connected to the other water territories
+              final SortedSet<String> waterTerrs2Copy = new TreeSet<String>(waterTerrs2);
+              for (Iterator<String> iter_waterTerr2 = waterTerrs2.iterator(); iter_waterTerr2.hasNext();) {
+                final String waterTerr2 = (String) iter_waterTerr2.next();
+                waterTerrs2Copy.remove(waterTerr2);
+                final SortedSet<String> waterTerrs2ReqNeighbors = new TreeSet<String>(waterTerrs2Copy);
+                final LinkedHashSet<String> waterTerr2Neightbors = MapXMLHelper.territoryConnections.get(waterTerr2);
+                if (waterTerr2Neightbors != null) {
+                  for (final String waterTerr2Neighbor : waterTerr2Neightbors)
+                    waterTerrs2ReqNeighbors.remove(waterTerr2Neighbor);
+                }
+                if (!waterTerrs2ReqNeighbors.isEmpty())
+                  iter_waterTerr2.remove();
+              }
+              // create canal only if at least 2 water territories remain

```

Comment:
quite a long method, anything over 50 lines is unlikely to get a thumbs up (short classes, short methods are preferred)